extension/flat_tensor/named_data_map/data_map.cpp

```diff
@@ -0,0 +1,215 @@
+      return fh.error();
+    }
+  }
+
+  // Load flatbuffer data as a segment.
+  Result<FreeableBuffer> flat_tensor_data = loader->load(
+      /*offset=*/flatbuffer_offset,
+      flatbuffer_size,
+      DataLoader::SegmentInfo(DataLoader::SegmentInfo::Type::External));
+  if (!flat_tensor_data.ok()) {
+    return flat_tensor_data.error();
+  }
+
+  // Make sure magic matches.
+  if (!flat_tensor_flatbuffer::FlatTensorBufferHasIdentifier(
+          flat_tensor_data->data())) {
+    ET_LOG(
+        Error,
+        "FlatTensor identifier '%.4s' != expected '%.4s'",
+        flatbuffers::GetBufferIdentifier(flat_tensor_data->data()),
+        flat_tensor_flatbuffer::FlatTensorIdentifier());
+    return Error::InvalidExternalData;
+  }
+
+  // The flatbuffer data must start at an aligned address to ensure internal
+  // alignment of flatbuffer fields.
+  ET_CHECK_OR_RETURN_ERROR(
+      IsAligned(flat_tensor_data->data()),
+      InvalidArgument,
+      "FlatTensor data 0x%p must be aligned to %zu",
+      flat_tensor_data->data(),
+      kMinimumAlignment);
+
+  // Get pointer to root of flatbuffer table.
+  const flat_tensor_flatbuffer::FlatTensor* flat_tensor =
+      flat_tensor_flatbuffer::GetFlatTensor(flat_tensor_data->data());
+
+  // Get pointer to tensor metadata.
+  const auto* s_tensor_metadata = flat_tensor->tensors();
+  assert(s_tensor_metadata != nullptr);
+
+  std::unordered_map<std::string, std::tuple<int, int, TensorLayout>>
+      name_to_tensor = {};
+  for (int i = 0; i < s_tensor_metadata->size(); i++) {
+    // Create TensorLayouts.
+    ScalarType scalar_type =
+        static_cast<ScalarType>(s_tensor_metadata->Get(i)->scalar_type());
+    const int dim = s_tensor_metadata->Get(i)->sizes()->size();
+
+    const auto serialized_sizes = s_tensor_metadata->Get(i)->sizes()->data();
+    const auto serialized_dim_order =
+        s_tensor_metadata->Get(i)->dim_order()->data();
+    TensorLayout tensor_layout = TensorLayout(
+        scalar_type,
+        Span<const int32_t>(serialized_sizes, dim),
+        Span<const uint8_t>(serialized_dim_order, dim));
+
+    int segment_index = s_tensor_metadata->Get(i)->segment_index();
+    int offset = s_tensor_metadata->Get(i)->offset();
+    std::string key = s_tensor_metadata->Get(i)->fully_qualified_name()->str();
+
+    auto val = std::make_tuple(segment_index, offset, tensor_layout);
+    name_to_tensor.insert({key, std::move(val)});
+  }
+
+  // Load constant data.
+  const auto* s_data_segment = flat_tensor->segments();
+
+  // Only support one segment for now.
+  assert(s_data_segment->size() == 1);
+  // First segment offset should be 0.
+  int segment_offset = s_data_segment->Get(0)->offset();
+  assert(segment_offset == 0);
+  // First segment size should be <= the total segment data size.
+  int segment_size = s_data_segment->Get(0)->size();
+  assert(segment_size <= segment_data_size);
+
+  Result<FreeableBuffer> _data_ro = loader->load(
```

Comment:
Ok I was trying to figure out how I wanted to handle this for mutable state. Here is what Ive come up with.

Lazily load segments, but internally keep a list of all the loaded segments. If the first time you need to load a segment is through "get_data" then put it in the preserve lis. If the first time you need to load a segment is through load_into then instead you just call data_loader load into which doesnt keep it hanging around.